kimchi/src/circuits/polynomials/foreign_field_mul/gadget.rs

```diff
@@ -0,0 +1,373 @@
+        witness,
+        joint_combiner,
+        table_id_combiner,
+        lookup_info,
+    )
+    .map_err(|_| CircuitGateError::InvalidLookupConstraintSorted(gate_type))?;
+
+    // Randomize the last `EVALS` rows in each of the sorted polynomials in order to add zero-knowledge to the protocol.
+    let sorted: Vec<_> = sorted
+        .into_iter()
+        .map(|chunk| lookup::constraints::zk_patch(chunk, cs.domain.d1, rng))
+        .collect();
+
+    let sorted_coeffs: Vec<_> = sorted.iter().map(|e| e.clone().interpolate()).collect();
+    let sorted8 = sorted_coeffs
+        .iter()
+        .map(|v| v.evaluate_over_domain_by_ref(cs.domain.d8))
+        .collect::<Vec<_>>();
+
+    // Compute the plookup aggregation evaluations
+    let aggreg = lookup::constraints::aggregation::<_, F>(
+        dummy_lookup_value,
+        &joint_lookup_table_d8,
+        cs.domain.d1,
+        &cs.gates,
+        witness,
+        &joint_combiner,
+        &table_id_combiner,
+        *beta,
+        *gamma,
+        &sorted,
+        rng,
+        lookup_info,
+    )
+    .map_err(|_| CircuitGateError::InvalidLookupConstraintAggregation(gate_type))?;
+
+    // Precompute different forms of the aggregation polynomial for later
+    let aggreg_coeffs = aggreg.interpolate();
+    // TODO: There's probably a clever way to expand the domain without interpolating
+    let aggreg8 = aggreg_coeffs.evaluate_over_domain_by_ref(cs.domain.d8);
+
+    Ok(LookupEnvironmentData {
+        aggreg8,
+        sorted8,
+        joint_lookup_table_d8,
+    })
+}
+
+// TODO: Check do we use this anywhere
+pub fn circuit_gate_selector_index(typ: GateType) -> usize {
+    match typ {
+        GateType::ForeignFieldMul => 0,
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get vector of foreign field multiplication circuit gate types
+pub fn circuit_gates() -> [GateType; GATE_COUNT] {
+    [GateType::ForeignFieldMul]
+}
+
+/// Get combined constraints for a given foreign field multiplication circuit gate
+pub fn circuit_gate_constraints<F: PrimeField>(typ: GateType, alphas: &Alphas<F>) -> E<F> {
+    match typ {
+        GateType::ForeignFieldMul => ForeignFieldMul::combined_constraints(alphas),
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Number of constraints for a given foreign field mul circuit gate type
+pub fn circuit_gate_constraint_count<F: PrimeField>(typ: GateType) -> u32 {
+    match typ {
+        GateType::ForeignFieldMul => ForeignFieldMul::<F>::CONSTRAINTS,
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get the combined constraints for all foreign field multiplication circuit gates
+pub fn combined_constraints<F: PrimeField>(alphas: &Alphas<F>) -> E<F> {
+    ForeignFieldMul::combined_constraints(alphas)
+}
+
+/// Get the foreign field multiplication lookup table
+pub fn lookup_table<F: PrimeField>() -> LookupTable<F> {
+    lookup::tables::get_table::<F>(GateLookupTable::RangeCheck)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Perhaps this is not where you wanted to have this test

Reply:
Even worse, this test doesn't even do anything.